### PR TITLE
Change approach for loading MIGraphX shared libs - append MIGRAPHX_SO_VERSION to the name

### PR DESCRIPTION
**Pull request by ahsan-ca**
This changes the way shared library names are used for loading the libraries. We default to loading library with so version appended to the name. If library with so version name is not found, then we try to  load the library without the so version name appended.

### CODE REVIEW

File: src/register_target.cpp [hunk 1]
```diff
@@ -57,7 +58,8 @@ target make_target(const std::string& name)
 {
     if(not contains(target_map(), name))
     {
-        auto target_name = make_shared_object_filename("migraphx_" + name);
+        auto target_name = make_shared_object_filename("migraphx_" + name) + "." +
+                           std::to_string(MIGRAPHX_SO_MAJOR_VERSION) + ".0";
```
Comment (pfultz2) [inline src/register_target.cpp:62]: First you need to try and load it with the version appended and if that fails then try it without the version.

File: src/dynamic_loader.cpp [hunk 2]
```diff
@@ -61,6 +62,24 @@ struct dynamic_loader_impl
...
+
+            // Erase so_version_name from the name of the library.
+            size_t pos = lib_name.find(so_version_name);
+            if(pos != std::string::npos)
+                lib_name.erase(pos, so_version_name.length());
+
+            // Load the library without the so version in the name.
+            handle = std::shared_ptr<void>(dlopen(lib_name.c_str(), RTLD_GLOBAL | RTLD_NOW),
+                                           manage_deleter<decltype(&dlclose), &dlclose>{});
+        }
```
Comment (pfultz2) [inline src/dynamic_loader.cpp:82]: This shouldn't be done here. Instead a try/catch should be used in register_target.cpp to check for both versions.

Comment (ahsan-ca) [reply src/dynamic_loader.cpp]: Done.

File: src/register_target.cpp [hunk 3]
```diff
@@ -57,8 +58,25 @@ target make_target(const std::string& name)
...
+        // try loading "libmigraphx_ref.so".
+        try
+        {
+            // Default to loading shared libraries with so version in their name.
+            store_target_lib(dynamic_loader(target_name + so_version));
+        }
+        catch(...)
+        {
+            // Load the library without the so version in the name.
+            std::cout << "In catch block..." << std::endl;
```
Comment (umangyadav) [inline src/register_target.cpp:77]: Remove this `cout`

File: src/version.h.in [hunk 4]
```diff
@@ -26,4 +26,8 @@
 #define MIGRAPHX_VERSION_MINOR @PROJECT_VERSION_MINOR@
 #define MIGRAPHX_VERSION_PATCH @PROJECT_VERSION_PATCH@
 #define MIGRAPHX_VERSION_TWEAK "@PROJECT_VERSION_TWEAK@"
+#define MIGRAPHX_SO_MAJOR_VERSION           \
```
Comment (umangyadav) [inline src/version.h.in:29]:
```suggestion
#define MIGRAPHX_SO_VERSION_FLAT           \
```
Suggestion:
I wouldn't put `major` in the word.  MIopen uses `FLAT` to represent such version string.  Perhaps good idea to use same naming convention.

File: src/register_target.cpp (hunk 3 above)
Comment (ahsan-ca) [reply src/register_target.cpp]: Good catch! I missed to remove my testing instructions. Thanks!

File: src/version.h.in (hunk 4 above)
Comment (ahsan-ca) [reply src/version.h.in:29]: CMakeLists.txt under AMDMIGraphX has this name, so I used that for consistency:
`math(EXPR MIGRAPHX_SO_MAJOR_VERSION "(${PROJECT_VERSION_MAJOR} * 1000 * 1000) + (${PROJECT_VERSION_MINOR} * 1000) + ${PROJECT_VERSION_PATCH}")`

I am open to changing the name if that is more desirable.

Comment (pfultz2) [reply src/version.h.in:29]: Yea this is the major version for the SO.

**Review by pfultz2 — APPROVED** [review]

**Review by umangyadav — APPROVED** [review]

**Pull request merged**